### PR TITLE
Initial AppleClang support

### PR DESCRIPTION
**Pull request by foxtran**
This patch fixes numerous of errors related to build of libra with AppleClang.

AppleClang lacks support of OpenMP, so, extra flag for disabling OpenMP can be passed. By default, OpenMP is enabled.

I have also updated CMakeLists, so currently, it requires at least CMake 3.12.

### CODE REVIEW

File: src/dyn/qtag/qtag.cpp [hunk 1]
```diff
@@ -593,7 +593,12 @@ complex<double> BATe(int i, int j,
 
 CMATRIX qtag_potential(MATRIX& q1, MATRIX& p1, MATRIX& s1, MATRIX& alp1, int n1, vector<int>& traj_on_surf_n1,
                        MATRIX& q2, MATRIX& p2, MATRIX& s2, MATRIX& alp2, int n2, vector<int>& traj_on_surf_n2,
-                       nHamiltonian& ham, int method, double AA, double BB, double CC){
+                       nHamiltonian& ham, int method, double ABC[3]){
```
Comment (foxtran) [inline src/dyn/qtag/qtag.cpp:596]: Boost::Python can not accept more than 16 args, so I collable three last into one. With simple code it works: https://godbolt.org/z/7qPxWEdbf

File: src/libint2_wrappers/libint2_wrappers.h [hunk 2]
```diff
@@ -82,7 +82,7 @@ void parallel_do(Lambda& lambda, int nthreads) {
 }
 #else  // use C++11 threads
 std::vector<std::thread> threads;
-for (int thread_id = 0; thread_id != libint2::nthreads; ++thread_id) {
+for (int thread_id = 0; thread_id != nthreads; ++thread_id) {
```
Comment (foxtran) [inline src/libint2_wrappers/libint2_wrappers.h:85]: I have no idea does libint2 namespace have nthreads variable or not. I suppose it is better to take the value from functions argument

**Review by foxtran — COMMENTED** [review]
Just some thought which I had during fixing compilation errors.

File: src/dyn/qtag/qtag.cpp (hunk 1 above)
Comment (alexvakimov) [reply src/dyn/qtag/qtag.cpp:596]: there is a macro - called something like ARITY - which enables lifting this restriction

File: src/dyn/dyn_hop_proposal.cpp [hunk 3]
```diff
@@ -1310,8 +1310,8 @@ int hop(vector<double>& prob, double ksi){
       } // for
 
     } // if nrm
-    else{  finstate = initstate; }  // probability to hop to any other states is zero
-                                    // so stay on the original state
+    else{  finstate = -1; }  // probability to hop to any other states is zero
+                             // so stay on the original state
```
[the remark this answers is not in the document]
Comment (alexvakimov) [reply src/dyn/dyn_hop_proposal.cpp:1314]: i think the `initstate` should be somewhere in the scope; setting it to -1 makes no sense other than for the purpose of spotting this situation with better probability; basically the initial guess is to set the final state to be the initial one. There shall be no negatively-indexed states

File: CMakeLists.txt [hunk 4]
```diff
@@ -24,34 +24,32 @@ LIST(APPEND CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake/modules)
...
-#FIND_PACKAGE(PythonLibs 3.7 REQUIRED)
-
-IF(PYTHONLIBS_FOUND)
-  MESSAGE("Success!")
-  INCLUDE_DIRECTORIES("${PYTHON_INCLUDE_DIRS}")
-ELSEIF(NOT PYTHONLIBS_FOUND)
-  MESSAGE("Unable to find correct Python version. If your Python include/libraries are installed in a non-standard location")
-  MESSAGE("Try using -DPYTHON_LIBRARY=\${path_to_the_python_library} -PYTHON_INCLUDE_DIR=\${path_to_where_Python.h_is_found}")
+MESSAGE(STATUS "Looking for Python libraries...")
+FIND_PACKAGE(Python3 3.6 REQUIRED COMPONENTS Interpreter Development)
```
Comment (alexvakimov) [inline CMakeLists.txt:42]: the version of Python can probably vary as long as that dev version of Python is installed; but i'd not be moving backwards to Python 3.6 in favor of 3.9 - why is that needed?

File: CMakeLists.txt [hunk 5]
```diff
@@ -60,66 +58,36 @@ MESSAGE(${PYTHON_INCLUDE_DIRS})
...
 SET(Boost_NO_BOOST_CMAKE TRUE) # to fix the problem of finding python37 libs
 SET(Boost_USE_STATIC_LIBS=OFF)
 SET(Boost_USE_MULTITHREADED=OFF)
 SET(Boost_USE_STATIC_RUNTIME=OFF)
 
-# Adapt it according to your system
-#FIND_PACKAGE(Boost 1.73.0 REQUIRED)
-FIND_PACKAGE(Boost 1.82.0 REQUIRED)
-#FIND_PACKAGE(Boost REQUIRED)
+FIND_PACKAGE(Boost 1.74.0 REQUIRED COMPONENTS regex python${LIBRA_Python3_VERSION})
```
Comment (alexvakimov) [inline CMakeLists.txt:67]: again, the latest version of Boost I could find at the moment is 1.82.0 - why to roll back to 1.74.0 version?
If this is a user's system limitation, this should be noted (e.g. users could adapt their CMakeLists file)

File: CMakeLists.txt [hunk 6]
```diff
@@ -60,66 +58,36 @@ MESSAGE(${PYTHON_INCLUDE_DIRS})
...
 IF(Boost_FOUND)
-  MESSAGE("Success!")
-  INCLUDE_DIRECTORIES("${Boost_INCLUDE_DIRS}") 
-  # Adapt it based on your system
-  #FIND_PACKAGE(Boost COMPONENTS python36 regex) # it seems that it is important to find sub-packages too!!!
-  FIND_PACKAGE(Boost COMPONENTS python39 regex)
```
Comment (alexvakimov) [inline CMakeLists.txt:78]: it is important to be able to find those Boost components so this can not be deleted

File: CMakeLists.txt [hunk 7]
```diff
@@ -60,66 +58,36 @@ MESSAGE(${PYTHON_INCLUDE_DIRS})
...
-  MESSAGE("${LIBINT2_INCLUDE_DIR}")
-  MESSAGE("Found Libint2 include directory: ")
-  MESSAGE("${Libint2_INCLUDE_DIRS}")
-  MESSAGE("Found Libint2_LIBRARIES: ")
-  MESSAGE("${Libint2_LIBRARIES}")
-  MESSAGE("Found Libint2_LIBRARY: ")
-  MESSAGE("${Libint2_LIBRARY}")
-ELSEIF(NOT Libint2_FOUND)
-  MESSAGE(FATAL_ERROR "Unable to find correct libint2")
-ENDIF()
```
[the remark this answers is not in the document]
Comment (alexvakimov) [reply CMakeLists.txt:122]: then why deleting them? maybe just keep them as they are for the time being; i feel like it likely will break something if this is accepted this way.

File: CMakeLists.txt [hunk 8]
```diff
@@ -139,18 +107,22 @@ ENDIF()
 #
 #  Cmake configuration
 #
-MESSAGE("Setting up the CMake configuration...")
+MESSAGE(STATUS "Setting up the CMake configuration...")
 IF(NOT CMAKE_BUILD_TYPE)
   SET(CMAKE_BUILD_TYPE "RELEASE")
-  SET(CMAKE_CXX_FLAGS "-fPIC -O2 -fpermissive -w -I ${Boost_INCLUDE_DIRS} -fopenmp")
```
Comment (alexvakimov) [inline CMakeLists.txt:145]: why to remove these flags?

Comment (alexvakimov) [reply CMakeLists.txt:145]: oh, i see you set them outside the if block; but again - the fpermissive was important at some point

File: src/dyn/heom/heom.cpp [hunk 9]
```diff
@@ -360,7 +360,7 @@ CMATRIX compute_deriv_n(int n, vector<CMATRIX>& rho, CMATRIX& Ham, vector<CMATRI
 
           scaling_factor = 1.0;
           if(do_scale==1){ // Scaled HEOM approach from JCP 130, 084105 (2009)
-            scaling_factor = sqrt((nvectors[n][m*(KK+1) + k] + 1.0 )*fabs(c_matsubara[k]));
+            scaling_factor = sqrt((nvectors[n][m*(KK+1) + k] + 1.0 )*std::abs(c_matsubara[k]));
```
[the remark this answers is not in the document]
Comment (alexvakimov) [reply src/dyn/heom/heom.cpp:363]: in this reference, both abs and fabs seem to be used interchangeably

File: src/dyn/qtag/CMakeLists.txt [hunk 10]
```diff
@@ -25,5 +25,7 @@ endif()
 #
 TARGET_LINK_LIBRARIES(qtag      linalg_stat ${ext_libs})
 TARGET_LINK_LIBRARIES(qtag_stat linalg_stat ${ext_libs})
+TARGET_LINK_LIBRARIES(qtag      nhamiltonian      ${ext_libs})
```
Comment (alexvakimov) [inline src/dyn/qtag/CMakeLists.txt:28]: just add "nhamiltonian_stat" (btw, I'm linking vs the static libraries, so it would be nhamiltonian_stat not nhamiltonian) in the list with "linalg_stat"  - why to separate into several lines?

File: src/dyn/qtag/libqtag.cpp [hunk 11]
```diff
@@ -66,7 +66,7 @@ void export_qtag_objects(){
   CMATRIX (*expt_qtag_potential_v1)
   (MATRIX& q1, MATRIX& p1, MATRIX& s1, MATRIX& alp1, int n1, vector<int>& traj_on_surf_n1,
    MATRIX& q2, MATRIX& p2, MATRIX& s2, MATRIX& alp2, int n2, vector<int>& traj_on_surf_n2,
-   nHamiltonian& ham, int method) = &qtag_potential;
+   nHamiltonian& ham, int method, double ABC[3]) = &qtag_potential;
```
Comment (alexvakimov) [inline src/dyn/qtag/libqtag.cpp:69]: we don't need to carry model parameters into this function anyways

File: src/libint2_wrappers/libint2_wrappers.h (hunk 2 above)
Comment (alexvakimov) [reply src/libint2_wrappers/libint2_wrappers.h:85]: i think using libint-namespace variables would probably be safer since this is something they use internally

File: src/math_linalg/base_matrix.h [hunk 12]
```diff
@@ -792,7 +792,7 @@ class base_matrix{
   ///< Assignment = copying (by value)
   void operator=(const base_matrix<T1>& ob){
 
-    if(this == &ob){  return *this;    }
```
Comment (alexvakimov) [inline src/math_linalg/base_matrix.h:795]: it should return *this - so that the comparison could be chained - this is as far as I understand the point of this return statement

File: src/models/Model_DAC.cpp [hunk 13]
```diff
@@ -27,7 +27,7 @@ namespace libmodels{
 
 
 void model_DAC(CMATRIX& Hdia, CMATRIX& Sdia, vector<CMATRIX>& d1ham_dia, vector<CMATRIX>& dc1_dia,
-               vector<double> q, vector<double>& params){ 
+               vector<double> &q, vector<double>& params){
```
Comment (alexvakimov) [inline src/models/Model_DAC.cpp:30]: agree here - passing by reference is better;

File: src/nhamiltonian/nHamiltonian_basic.cpp [hunk 14]
```diff
@@ -310,7 +310,7 @@ void nHamiltonian::copy_content(nHamiltonian* src){
 
 void nHamiltonian::copy_content(const nHamiltonian& src){
 
-  copy_content(&src);
```
Comment (alexvakimov) [inline src/nhamiltonian/nHamiltonian_basic.cpp:313]: should be the same - no? also make sure this doesn't mess up any other places

**Review by alexvakimov — CHANGES_REQUESTED** [review]

**Comment by alexvakimov** [comment]
Btw, I really appreciate your contributions, Igor @foxtran , but can you make a separate branch - i'm a bit scared to accept the current changes to the devel one. Not sure it it would all compile for us. 
Also, have you tested the build for the clang? Perhaps adding some comments for the Mac users - on how to setup the environment and build the code would be really useful.

File: src/dyn/heom/heom.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (foxtran) [reply src/dyn/heom/heom.cpp:363]: fabs is only for real variables. See more here: https://en.cppreference.com/w/cpp/numeric/math/fabs

File: src/dyn/qtag/qtag.cpp (hunk 1 above)
Comment (foxtran) [reply src/dyn/qtag/qtag.cpp:596]: Hmmm... But somewhat it did not work for me

File: CMakeLists.txt (hunk 4 above)
Comment (foxtran) [reply CMakeLists.txt:42]: Find_Package looks for the version, which is not lower than defined. I have installed Python3.9, 3.10, 3.11 and 3.12. The latest one was chosen.

File: CMakeLists.txt (hunk 5 above)
Comment (foxtran) [reply CMakeLists.txt:67]: It is a minimal supported version. I have this version. If you have 1.82, it will be also found.

File: CMakeLists.txt (hunk 6 above)
Comment (foxtran) [reply CMakeLists.txt:78]: See line with find_package statement: it will look for theses components.

File: CMakeLists.txt (hunk 8 above)
Comment (foxtran) [reply CMakeLists.txt:145]: -fpermissive is not a safe option. Better to avoid it. Looks like it was a main issue with AppleClang compiler

File: src/dyn/qtag/libqtag.cpp (hunk 11 above)
Comment (foxtran) [reply src/dyn/qtag/libqtag.cpp:69]: Then, what is expected as these model parameters? I did not notice default values...

**Comment by foxtran** [comment]
Ok. I will leave only source code changes, and return tomorrow with another PR :)

**Comment by alexvakimov** [comment]
I actually, like some of the CMakeLists changes you did, @foxtran , but it would be better to have either a separate branch or another CMakeLists file (so the users could chose the one that suits them most)

File: src/math_linalg/base_matrix.h (hunk 12 above)
Comment (foxtran) [reply src/math_linalg/base_matrix.h:795]: This function is marked as void, so, it cannot return anything

Comment (foxtran) [reply src/math_linalg/base_matrix.h:795]: Hmm... You are right, the another patch should be here

File: src/dyn/dyn_hop_proposal.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (foxtran) [reply src/dyn/dyn_hop_proposal.cpp:1314]: It is not in the scope. After some looking, I have found the following calls:
https://github.com/Quantum-Dynamics-Hub/libra-code/blob/14a99c738ef09d9a58ccdc83aeab611e72bbc2e1/src/dyn/dyn_projectors.cpp#L296
https://github.com/Quantum-Dynamics-Hub/libra-code/blob/14a99c738ef09d9a58ccdc83aeab611e72bbc2e1/src/dyn/dyn_projectors.cpp#L528

`hop(int, std::vector<double> &, double)` has exactly the same code, as `hop(std::vector<double> &, double)`, except value of initstate.

File: src/converters/converters.h [hunk 15]
```diff
@@ -95,19 +95,19 @@ template<class T>
 struct std_item
 {
     typedef typename T::value_type V;
-    static V& get(T const& x, int i)
+    static V& get(T & x, int i)
```
Comment (alexvakimov) [inline src/converters/converters.h:98]: these changes would be conflicting with those in your other PR

**Review by alexvakimov — COMMENTED** [review]
Okay, looks like your other PR implied you don't want to keep this PR (there are some overlaps with the changes in this PR). Let me close this one and we focus on the other one. I'll experiment with your CMakeLists suggestions and will incorporate them if they don't break the gcc build. I'll close this PR for now.
